Add product deletion via datastore API

octaedro opened the pull request:
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR adds product deletion via the datastore API.

Closes #33127.

### How to test the changes in this Pull Request:

1. Go to `wp-admin/edit.php?post_type=product`
2. Pick an ID of a recently published product
3. Open your browser’s console
4. Run the following.  You may need to run again after the fetch has been completed.
```
5. wp.data.select('wc/admin/products').getProducts({ status: 'publish'}, undefined);
```
6. Note the product you picked.
7. Enter the following, replacing the `my_product_id` ID with your product ID:
```
wp.data.dispatch('wc/admin/products').removeProduct( `my_product_id` );
```
8. Run the getProducts selector again:
```
wp.data.select('wc/admin/products').getProducts({ status: 'publish'}, undefined);
```
9. Note the product status now is `trash`.
10. Go to the `Trash` can (`/wp-admin/edit.php?post_status=trash&post_type=product`) and verify the product is there.
11. Now run the following line to delete the product definitively
```
wp.data.dispatch('wc/admin/products').removeProduct( `my_product_id`, { force: true } );
```
12. Run the getProducts selector again and note the product status now is `deleted`.
13. Go to the product list and the `Trash` can and verify the product is not present anymore.


### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [x] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm nx changelog <project>`?

<!-- Mark completed items with an [x] -->

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

botwoo commented:
:bar_chart: Test reports for this pull request have been published and are accessible through the following links:

- [REST API Test Report](https://woocommerce.github.io/woocommerce-test-reports/pr/33285/api/)
- E2E Test Report (TBC)

Latest commit referenced in the reports: `Add product deletion via datastore API` 3b2a07543657bd9dbddd29bebdde5f533fb24f95
This comment will automatically be updated with the latest referenced commit when you push new changes to this pull request.

---

_Visit the [WooCommerce Test Reports](https://woocommerce.github.io/woocommerce-test-reports) homepage to view all published reports. See the [FAQs](https://github.com/woocommerce/woocommerce-test-reports/wiki/FAQs) page if you're having problems accessing them._

octaedro committed:
Add delete products via data store api

# Conflicts:
#	packages/js/data/src/products/action-types.ts
#	packages/js/data/src/products/actions.ts
#	packages/js/data/src/products/reducer.ts
#	packages/js/data/src/products/selectors.ts

octaedro committed:
Add tests

# Conflicts:
#	packages/js/data/src/products/test/reducer.ts

octaedro committed:
Set `force = false` by default

octaedro committed:
Add changelog

joshuatf commented on packages/js/data/src/products/action-types.ts:
```diff
@@ -7,6 +7,8 @@ export enum TYPES {
 	GET_PRODUCTS_ERROR = 'GET_PRODUCTS_ERROR',
 	GET_PRODUCTS_TOTAL_COUNT_SUCCESS = 'GET_PRODUCTS_TOTAL_COUNT_SUCCESS',
 	GET_PRODUCTS_TOTAL_COUNT_ERROR = 'GET_PRODUCTS_TOTAL_COUNT_ERROR',
+	DELETE_PRODUCT_SUCCESS = 'DELETE_PRODUCT_SUCCESS',
```
Nice!  Thanks for using the naming conventions 😄

joshuatf commented on packages/js/data/src/products/actions.ts:
```diff
@@ -116,6 +117,54 @@ export function* createProduct( data: Pick< Product, MutableProperties > ) {
+
+export function* removeProduct(
+	id: number,
+	type: DeleteProductType = { force: false }
```
Will there be more types on this in the future?  If not, should `type` just be changed to `force` as a simple boolean?

octaedro replied:
Good idea! I was trying to have something flexible for the future, but it definitely was an early optimization. I fixed it in commit 5aa5540

joshuatf commented on packages/js/data/src/products/actions.ts:
```diff
@@ -116,6 +117,54 @@ export function* createProduct( data: Pick< Product, MutableProperties > ) {
+		throw new Error();
+	} catch ( error ) {
+		yield deleteProductError( id, error );
+		throw new Error();
```
We should probably throw the `error` from above here so it can be used by the consumer.

octaedro replied:
Sounds good to me. I added that change in commit 11cde28

joshuatf commented on packages/js/data/src/products/actions.ts:
```diff
@@ -116,6 +117,54 @@ export function* createProduct( data: Pick< Product, MutableProperties > ) {
+			return product;
+		}
+
+		throw new Error();
```
When does this happen?  Is it when a product could not be successfully deleted?  I'm wondering when `catch` would not work and if this can happen, maybe we should add some message to the error about how/why it would happen.

octaedro replied:
Nice catch! I removed that in commit 63b9e7e

joshuatf commented on packages/js/data/src/products/reducer.ts:
```diff
@@ -99,6 +99,27 @@ const reducer: Reducer< ProductState, Actions > = (
+				const prData = state.data || {};
+				return {
+					...state,
+					data: {
```
Do we also need to remove the deleted from various `products: { [itemType]: [...] }` arrays?

At a quick glance it seems like we could handle this one of two ways:
1. The item gets removed from stores.  This might be problematic since a query could be specifically looking for products with `trash` status.
2. Leave it as is and let the consumer filter out deleted products.

Not really sure what the best method is, but I'm leaning towards option 2; just wanted to get some discussion around this.

joshuatf replied:
Thinking about this more, anytime we update or delete a product, it may no longer belong to the query resource name it was previously assigned.

cc @louwie17 for your thoughts on this as well.

louwie17 replied:
Yeah, I have a couple thoughts:
- For deleted products we could use the filter either where we display the list of products or within the selector itself, we did probably want to create a different selector `getProductsWithStatus` in that case, that makes use of `getProducts` under the hood. The [use of `useSelector`](https://github.com/woocommerce/woocommerce/blob/trunk/packages/js/data/src/products/selectors.ts#L51-L55) , does allow the component to be re-rendered if a nested product has been updated (whether that is it's status, title, or anything else). So as long we keep any other filters client side, we should be fine on that front.
 
- Make use of `invalidateResolution`, probably only from the component the deletion or update takes place in and the query as well. I am not sure if there is an easy way for us to make this part of the `delete` and `update` actions. I guess we could query for the specific `id` within `products[queryString].data` and invalidate all of them that contain the `id`. As long as we do this silently I think it is fine (meaning we don't wipe the entire table and show it is refreshing).

- Change our redux store, by removing the `products[queryString].data` part and only having a single list of products. We did have to do some filtering on the client side also. This would cause an issue if we ever wanted to display multiple lists of products on the same page (unless we filter everything, including pagination on the front-end).

I say a combination of the first two should be fine.

louwie17 replied:
> Leave it as is and let the consumer filter out deleted products.

I am really thinking that this would be the way to go, it all depends on what our layout will look like for a list of products.

octaedro replied:
Thank you @joshuatf and @louwie17 for your feedback on this. So if it's ok with you I will leave the current handling of the deleted products like it is, to continue analyzing the resolution invalidation and tackle it in a follow-up PR.

joshuatf reviewed (commented):
Nice work, @octaedro!  Just left a few questions about how we handle things.

There's one possibly bigger issue with how we keep track of previously fetched product queries, but I'm happy to leave this for another PR if it turns out to be a big issue.

octaedro committed:
Remove error

octaedro committed:
Remove `DeleteProductType`

# Conflicts:
#	packages/js/data/src/products/actions.ts

octaedro committed:
Fix thrown error

octaedro commented:
Hey @joshuatf,
I addressed the changes you suggested. Could you take another look at this PR?

joshuatf commented on packages/js/data/src/products/actions.ts:
```diff
@@ -151,6 +151,48 @@ export function* updateProduct(
+			method: 'DELETE',
+		} );
+
+		if ( product ) {
```
Any point in wrapping this in an "if" still?  I would assume we always want to return whatever is returned from the response if it's a `200`.

octaedro replied:
Good idea @joshuatf! I changed it in the commit 5121432
I think that this PR should be reapproved. Could you take another look at this PR?

joshuatf reviewed (dismissed):
Testing well!  Left one optional comment, but otherwise code looks great and I'm able to delete and force delete products without issue!

octaedro committed:
Remove conditional retrun

joshuatf reviewed (approved):
Nice!  Thanks for that last change, Fernando!  LGTM 🚢
